spark-3.1-spanner-lib/src/test/java/com/google/cloud/spark/spanner/SpannerTestBase.java

```diff
@@ -121,6 +121,45 @@ private static void initDatabase() throws Exception {
 
               return null;
             });
+
+    // 3.2. Insert the Shakespeare data.
+    int maxValuesPerTxn = 35_000;
+    int MAX_BYTE_SIZE_PER_TXN = 1_048_576;
+
+    List<String> shakespearValues = TestData.shakespearValues;
+    for (int start = 0; start < shakespearValues.size(); ) {
+      int end = start + maxValuesPerTxn;
+      if (end > shakespearValues.size()) {
+        end = shakespearValues.size();
+      }
+
+      String sqlPrefix =
+          "INSERT INTO Shakespeare(id, word, word_count, corpus, corpus_date) VALUES";
+      int i = start;
+      int byteSize = sqlPrefix.length();
```

Comment:
You can partition the list by using: 
Lists.partition(list, chunkSize);

https://www.geeksforgeeks.org/java-guava-lists-partition-method-with-examples/

Reply:
So there are 2 constraints:
1. Max of 40,000 statements per transaction
2. Splitting by max byte size per commit: 1MiB

The current code suffices as we go through both constraints.

Reply:
Okay, it's your decision. 

What I proposed is just making the code simpler. Something like 

Lists.partition(shakespearValues, 1000); could just return List<List<String>> 

then you can just iterate the operation on sub partitions.